File: artifacts/genann/refactor/example3.c
#include <example3.c>
#include <genann.h>
#include <stdio.h>
#include <stdlib.h>

struct genann;
typedef double (*genann_actfun)(const struct genann *ann, double a);
typedef struct genann
{
  int inputs;
  int hidden_layers;
  int hidden;
  int outputs;
  genann_actfun activation_hidden;
  genann_actfun activation_output;
  int total_weights;
  int total_neurons;
  double *weight;
  double *output;
  double *delta;
} genann;
genann *genann_init(int inputs, int hidden_layers, int hidden, int outputs);
genann *genann_read(FILE *in);
void genann_randomize(genann *ann);
genann *genann_copy(const genann *ann);
void genann_free(genann *ann);
const double *genann_run(const genann *ann, const double *inputs);
void genann_train(const genann *ann, const double *inputs, const double *desired_outputs, double learning_rate);
void genann_write(const genann *ann, FILE *out);
void genann_init_sigmoid_lookup(const genann *ann);
double genann_act_sigmoid(const genann *ann, double a);
double genann_act_sigmoid_cached(const genann *ann, double a);
double genann_act_threshold(const genann *ann, double a);
double genann_act_linear(const genann *ann, double a);
const char *save_name = "example/xor.ann";
int main(int argc, char *argv[])
{
  printf("GENANN example 3.\n");
  printf("Load a saved ANN to solve the XOR function.\n");
  FILE *saved = fopen(save_name, "r");
  if (!saved)
  {
    printf("Couldn't open file: %s\n", save_name);
    exit(1);
  }
  genann *ann = genann_read(saved);
  fclose(saved);
  if (!ann)
  {
    printf("Error loading ANN from file: %s.", save_name);
    exit(1);
  }
  const double input[4][2] = {{0, 0}, {0, 1}, {1, 0}, {1, 1}};
  printf("Output for [%1.f, %1.f] is %1.f.\n", input[0][0], input[0][1], *genann_run(ann, input[0]));
  printf("Output for [%1.f, %1.f] is %1.f.\n", input[1][0], input[1][1], *genann_run(ann, input[1]));
  printf("Output for [%1.f, %1.f] is %1.f.\n", input[2][0], input[2][1], *genann_run(ann, input[2]));
  printf("Output for [%1.f, %1.f] is %1.f.\n", input[3][0], input[3][1], *genann_run(ann, input[3]));
  genann_free(ann);
  return 0;
}


struct genann;
typedef double (*genann_actfun)(const struct genann *ann, double a);
typedef struct genann
{
  int inputs;
  int hidden_layers;
  int hidden;
  int outputs;
  genann_actfun activation_hidden;
  genann_actfun activation_output;
  int total_weights;
  int total_neurons;
  double *weight;
  double *output;
  double *delta;
} genann;
genann *genann_init(int inputs, int hidden_layers, int hidden, int outputs);
genann *genann_read(FILE *in);
void genann_randomize(genann *ann);
genann *genann_copy(const genann *ann);
void genann_free(genann *ann);
const double *genann_run(const genann *ann, const double *inputs);
void genann_train(const genann *ann, const double *inputs, const double *desired_outputs, double learning_rate);
void genann_write(const genann *ann, FILE *out);
void genann_init_sigmoid_lookup(const genann *ann);
double genann_act_sigmoid(const genann *ann, double a);
double genann_act_sigmoid_cached(const genann *ann, double a);
double genann_act_threshold(const genann *ann, double a);
double genann_act_linear(const genann *ann, double a);
const char *save_name = "example/xor.ann";
int main(int argc, char *argv[])
{
  printf("GENANN example 3.\n");
  printf("Load a saved ANN to solve the XOR function.\n");
  FILE *saved = fopen(save_name, "r");
  if (!saved)
  {
    printf("Couldn't open file: %s\n", save_name);
    exit(1);
  }
  genann *ann = genann_read(saved);
  fclose(saved);
  if (!ann)
  {
    printf("Error loading ANN from file: %s.", save_name);
    exit(1);
  }
  const double input[4][2] = {{0, 0}, {0, 1}, {1, 0}, {1, 1}};
  printf("Output for [%1.f, %1.f] is %1.f.\n", input[0][0], input[0][1], *genann_run(ann, input[0]));
  printf("Output for [%1.f, %1.f] is %1.f.\n", input[1][0], input[1][1], *genann_run(ann, input[1]));
  printf("Output for [%1.f, %1.f] is %1.f.\n", input[2][0], input[2][1], *genann_run(ann, input[2]));
  printf("Output for [%1.f, %1.f] is %1.f.\n", input[3][0], input[3][1], *genann_run(ann, input[3]));
  genann_free(ann);
  return 0;
}

File: artifacts/genann/translate_new_ablation/src/example3.rs
use std::env;
use std::fs::File;
use std::io::Read;
use std::io;
use std::process::exit;
use std::process;
use translate_without_break_down::*;
pub fn main() {
    println!("GENANN example 3.");
    println!("Load a saved ANN to solve the XOR function.");
    
    let mut saved = match File::open(SAVE_NAME) {
        Ok(file) => file,
        Err(_) => {
            println!("Couldn't open file: {}", SAVE_NAME);
            exit(1);
        }
    };
    
    let mut ann = genann_read(&mut saved);
    if ann.is_none() {
        println!("Error loading ANN from file: {}.", SAVE_NAME);
        exit(1);
    }
    
    let input = [[0.0, 0.0], [0.0, 1.0], [1.0, 0.0], [1.0, 1.0]];
    
    if let Some(output) = genann_run(ann.as_ref().unwrap(), &input[0]) {
        if let Some(first_output) = output.first() {
            println!("Output for [{:.1}, {:.1}] is {:.1}.", input[0][0], input[0][1], first_output);
        }
    }
    
    if let Some(output) = genann_run(ann.as_ref().unwrap(), &input[1]) {
        if let Some(first_output) = output.first() {
            println!("Output for [{:.1}, {:.1}] is {:.1}.", input[1][0], input[1][1], first_output);
        }
    }
    
    if let Some(output) = genann_run(ann.as_ref().unwrap(), &input[2]) {
        if let Some(first_output) = output.first() {
            println!("Output for [{:.1}, {:.1}] is {:.1}.", input[2][0], input[2][1], first_output);
        }
    }
    
    if let Some(output) = genann_run(ann.as_ref().unwrap(), &input[3]) {
        if let Some(first_output) = output.first() {
            println!("Output for [{:.1}, {:.1}] is {:.1}.", input[3][0], input[3][1], first_output);
        }
    }
    
    genann_free(ann.map(Box::new));
}
